File: scripts/port263/census.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Census helper for the MC 26.3 port.

Runs javac over the whole source set with the *real* NeoForge/MC compile
classpath (materialised by `gradlew dumpCompileClasspath` into
build/compile-cp.txt) and prints an aggregated error breakdown.

javac stops at 100 errors by default, which is useless for a port of this size,
so -Xmaxerrs is raised. Gradle's own output is truncated by the harness, which
is the other reason this runs javac directly: we need the complete list to
drive codemods instead of fixing errors 100 at a time.

Usage:  python scripts/port263/census.py [--out .ref263/census.txt]
"""
import collections
import io
import json
import os
import re
import subprocess
import sys

ROOT = os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..'))
JAVA_HOME = r'D:\MC\jdk\graalvm-25.2.4+7.1'
CENSUS_DIR = os.path.join(ROOT, '.ref263')
HEADER = re.compile(r'^(.*?\.java):(\d+): error: (.*)$')


def compile_all(out_path):
    cp_file = os.path.join(ROOT, 'build', 'compile-cp.txt')
    if not os.path.exists(cp_file):
        sys.exit('missing build/compile-cp.txt - run: gradlew dumpCompileClasspath')
    cp = io.open(cp_file, encoding='utf-8').read().strip()

    sources = []
    for dp, _, fs in os.walk(os.path.join(ROOT, 'src', 'main', 'java')):
        sources += [os.path.join(dp, f) for f in fs if f.endswith('.java')]
    io.open(os.path.join(CENSUS_DIR, 'sources.txt'), 'w', encoding='utf-8', newline='\n').write(
        '\n'.join(sources))
    os.makedirs(os.path.join(CENSUS_DIR, 'classes'), exist_ok=True)

    cmd = [os.path.join(JAVA_HOME, 'bin', 'javac.exe'), '-J-Duser.language=en',
           '-Xmaxerrs', '30000', '-nowarn', '-proc:none', '-encoding', 'UTF-8',
           '-d', os.path.join(CENSUS_DIR, 'classes'), '-cp', cp,
           '@' + os.path.join(CENSUS_DIR, 'sources.txt')]
    r = subprocess.run(cmd, capture_output=True, text=True, errors='replace')
    io.open(out_path, 'w', encoding='utf-8', newline='\n').write(r.stdout + r.stderr)
    return len(sources)


def parse(path):
    errors = []
    cur = None
    for line in io.open(path, encoding='utf-8', errors='replace').read().splitlines():
        m = HEADER.match(line)
        if m:
            cur = {'file': m.group(1), 'line': int(m.group(2)), 'msg': m.group(3),
                   'symbol': None, 'loc': None}
            errors.append(cur)
        elif cur is not None:
            s = line.strip()
            if s.startswith('symbol:'):
                cur['symbol'] = s[7:].strip()
            elif s.startswith('location:'):
                cur['loc'] = s[9:].strip()
    return errors


def main():
    out = sys.argv[sys.argv.index('--out') + 1] if '--out' in sys.argv else \
        os.path.join(CENSUS_DIR, 'census.txt')
    n = compile_all(out)
    errors = parse(out)
    print('sources: %d   ERRORS: %d   files affected: %d' % (
        n, len(errors), len({e['file'] for e in errors})))
    print('\n=== TOP MESSAGES ===')
    for msg, c in collections.Counter(e['msg'] for e in errors).most_common(20):
        print('%6d  %s' % (c, msg))
    print('\n=== TOP MISSING SYMBOLS ===')
    sym = collections.Counter()
    for e in errors:
        if 'cannot find symbol' in e['msg'] and e['symbol']:
            sym[re.sub(r'^(class|variable|method|interface|enum)\s+', '', e['symbol'])] += 1
    for s, c in sym.most_common(40):
        print('%6d  %s' % (c, s))
    print('\n=== WORST FILES ===')
    for f, c in collections.Counter(e['file'] for e in errors).most_common(20):
        print('%6d  %s' % (c, os.path.relpath(f, ROOT)))
    json.dump(errors, io.open(out.replace('.txt', '.json'), 'w', encoding='utf-8'),
              ensure_ascii=False)


if __name__ == '__main__':
    main()

File: scripts/port263/census.mjs
#!/usr/bin/env node
// Census helper for the MC 26.3 port (Node port of scripts/port263/census.py:
// this machine has no python interpreter).
//
// Runs javac over the whole source set with the *real* NeoForge/MC compile
// classpath (materialised by `gradlew dumpCompileClasspath` into
// build/compile-cp.txt) and prints an aggregated error breakdown.
//
// javac stops at 100 errors by default, which is useless for a port of this size,
// so -Xmaxerrs is raised. Gradle's own output is truncated by the harness, which
// is the other reason this runs javac directly: we need the complete list to
// drive codemods instead of fixing errors 100 at a time.
//
// Usage:  node scripts/port263/census.mjs [--out .ref263/censusN.txt]

import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
const JAVA_HOME = process.env.CSRP_JAVA_HOME ?? "D:\\MC\\jdk\\graalvm-25.2.4+7.1";
const CENSUS_DIR = path.join(ROOT, ".ref263");
const HEADER = /^(.*?\.java):(\d+): error: (.*)$/;

function walk(dir, out = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) walk(full, out);
    else if (entry.name.endsWith(".java")) out.push(full);
  }
  return out;
}

function compileAll(outPath) {
  const cpFile = path.join(ROOT, "build", "compile-cp.txt");
  if (!fs.existsSync(cpFile)) {
    console.error("missing build/compile-cp.txt - run: gradlew dumpCompileClasspath");
    process.exit(2);
  }
  const cp = fs.readFileSync(cpFile, "utf8").trim();
  const sources = walk(path.join(ROOT, "src", "main", "java"));
  const argFile = path.join(CENSUS_DIR, "sources.txt");
  fs.writeFileSync(argFile, sources.join("\n"), "utf8");
  fs.mkdirSync(path.join(CENSUS_DIR, "classes"), { recursive: true });

  const javac = path.join(JAVA_HOME, "bin", "javac.exe");
  const args = [
    "-J-Duser.language=en", "-J-Duser.country=US",
    // javac writes diagnostics in the *console* charset unless told otherwise, which turns the
    // workspace path (D:\code\MC模组\...) into mojibake and breaks every later source lookup.
    "-J-Dfile.encoding=UTF-8", "-J-Dstdout.encoding=UTF-8", "-J-Dstderr.encoding=UTF-8",
    "-Xmaxerrs", "30000", "-nowarn", "-proc:none", "-encoding", "UTF-8",
    "-d", path.join(CENSUS_DIR, "classes"), "-cp", cp, `@${argFile}`
  ];
  const r = spawnSync(javac, args, { encoding: "buffer", maxBuffer: 1 << 30 });
  if (r.error) {
    console.error(`failed to run javac: ${r.error.message}`);
    process.exit(2);
  }
  const bytes = Buffer.concat([r.stdout ?? Buffer.alloc(0), r.stderr ?? Buffer.alloc(0)]);
  let text = bytes.toString("utf8");
  if (text.includes("\uFFFD")) {
    // Fall back to the legacy console charset if javac ignored the encoding properties.
    try {
      text = new TextDecoder("gbk").decode(bytes);
    } catch {
      // keep the UTF-8 best effort
    }
  }
  fs.writeFileSync(outPath, text, "utf8");
  return sources.length;
}

function parse(text) {
  const errors = [];
  let cur = null;
  for (const line of text.split(/\r?\n/)) {
    const m = HEADER.exec(line);
    if (m) {
      cur = { file: m[1], line: Number(m[2]), msg: m[3], symbol: null, loc: null };
      errors.push(cur);
      continue;
    }
    if (!cur) continue;
    const s = line.trim();
    if (s.startsWith("symbol:")) cur.symbol = s.slice(7).trim();
    else if (s.startsWith("location:")) cur.loc = s.slice(9).trim();
  }
  return errors;
}

function count(values) {
  const map = new Map();
  for (const v of values) map.set(v, (map.get(v) ?? 0) + 1);
  return [...map.entries()].sort((a, b) => b[1] - a[1]);
}

function main() {
  const outIndex = process.argv.indexOf("--out");
  const out = outIndex >= 0 ? process.argv[outIndex + 1] : path.join(CENSUS_DIR, "census.txt");
  const sourceCount = compileAll(out);
  const text = fs.readFileSync(out, "utf8");
  const errors = parse(text);

  console.log(`sources: ${sourceCount}   ERRORS: ${errors.length}   files affected: ${
    new Set(errors.map((e) => e.file)).size}`);
  console.log("\n=== TOP MESSAGES ===");
  for (const [msg, c] of count(errors.map((e) => e.msg)).slice(0, 20)) {
    console.log(`${String(c).padStart(6)}  ${msg}`);
  }
  console.log("\n=== TOP MISSING SYMBOLS ===");
  const symbols = errors
    .filter((e) => e.msg.includes("cannot find symbol") && e.symbol)
    .map((e) => e.symbol.replace(/^(class|variable|method|interface|enum)\s+/, ""));
  for (const [sym, c] of count(symbols).slice(0, 40)) {
    console.log(`${String(c).padStart(6)}  ${sym}`);
  }
  console.log("\n=== WORST FILES ===");
  for (const [file, c] of count(errors.map((e) => e.file)).slice(0, 20)) {
    console.log(`${String(c).padStart(6)}  ${path.relative(ROOT, file)}`);
  }
  fs.writeFileSync(out.replace(/\.txt$/, ".json"), JSON.stringify(errors), "utf8");
}

main();
